PostProcess/Common/src/PostFXContext.cpp

```diff
@@ -0,0 +1,226 @@
+{
+    DEV_CHECK_ERR(RenderAttribs.pDevice != nullptr, "RenderAttribs.pDevice must not be null");
+    DEV_CHECK_ERR(RenderAttribs.pDeviceContext != nullptr, "RenderAttribs.pDeviceContext must not be null");
+
+    ScopedDebugGroup DebugGroupGlobal{RenderAttribs.pDeviceContext, "PreparePostFX"};
+
+    if (RenderAttribs.pBufferCameraAttribs == nullptr)
+    {
+        if (!m_Resources[RESOURCE_IDENTIFIER_CONSTANT_BUFFER])
+        {
+            RefCntAutoPtr<IBuffer> pBuffer;
+            CreateUniformBuffer(RenderAttribs.pDevice, sizeof(HLSL::CombineCameraAttribs), "PostFXContext::ConstantBuffer", &pBuffer);
+            m_Resources[RESOURCE_IDENTIFIER_CONSTANT_BUFFER] = pBuffer;
+        }
+
+        MapHelper<HLSL::CombineCameraAttribs> CameraAttibs(RenderAttribs.pDeviceContext, reinterpret_cast<IBuffer*>(m_Resources[RESOURCE_IDENTIFIER_CONSTANT_BUFFER].RawPtr()), MAP_WRITE, MAP_FLAG_DISCARD);
+        *CameraAttibs                        = *RenderAttribs.pCameraAttribs;
+        CameraAttibs->CurrCamera.iFrameIndex = RenderAttribs.FrameIndex;
+    }
+    else
+    {
+        m_Resources[RESOURCE_IDENTIFIER_CONSTANT_BUFFER] = RenderAttribs.pBufferCameraAttribs;
+    }
+
+    m_CurrentFrameIndex = RenderAttribs.FrameIndex;
+
+    auto& RenderTech = m_RenderTech[RENDER_TECH_COMPUTE_BLUE_NOISE_TEXTURE];
+    if (!RenderTech.IsInitialized())
+    {
+        PipelineResourceLayoutDescX ResourceLayout;
+        ResourceLayout
+            .AddVariable(SHADER_TYPE_PIXEL, "cbCameraAttribs", SHADER_RESOURCE_VARIABLE_TYPE_STATIC)
+            .AddVariable(SHADER_TYPE_PIXEL, "g_SobolBuffer", SHADER_RESOURCE_VARIABLE_TYPE_STATIC)
+            .AddVariable(SHADER_TYPE_PIXEL, "g_ScramblingTileBuffer", SHADER_RESOURCE_VARIABLE_TYPE_STATIC);
+
+        const auto VS = PostFXRenderTechnique::CreateShader(RenderAttribs.pDevice, RenderAttribs.pStateCache, "FullScreenTriangleVS.fx", "FullScreenTriangleVS", SHADER_TYPE_VERTEX);
+        const auto PS = PostFXRenderTechnique::CreateShader(RenderAttribs.pDevice, RenderAttribs.pStateCache, "ComputeBlueNoiseTexture.fx", "ComputeBlueNoiseTexturePS", SHADER_TYPE_PIXEL);
+
+        RenderTech.InitializePSO(RenderAttribs.pDevice,
+                                 RenderAttribs.pStateCache, "PreparePostFX::ComputeBlueNoiseTexture",
+                                 VS, PS, ResourceLayout,
+                                 {
+                                     GetInternalResourceFormat(m_Resources[RESOURCE_IDENTIFIER_BLUE_NOISE_TEXTURE_XY]),
+                                     GetInternalResourceFormat(m_Resources[RESOURCE_IDENTIFIER_BLUE_NOISE_TEXTURE_ZW]),
+                                 },
+                                 TEX_FORMAT_UNKNOWN,
+                                 DSS_DisableDepth, BS_Default, false);
+        RenderTech.PSO->GetStaticVariableByName(SHADER_TYPE_PIXEL, "cbCameraAttribs")->Set(m_Resources[RESOURCE_IDENTIFIER_CONSTANT_BUFFER]);
```

Comment:
I recently added new `ShaderResourceVariableX` class, which is more safe, so let's use it:

```
ShaderResourceVariableX{RenderTech.PSO, SHADER_TYPE_PIXEL, "cbCameraAttribs"}.Set(m_Resources[RESOURCE_IDENTIFIER_CONSTANT_BUFFER])
```

Reply:
It is also worth allowing render techniques to store variables.

Reply:
Talked about it on Discord